File: Software Engineering/Computer Science/Lab2/HammingCode.java
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
public class HammingCode {
    public static void main(String[] args) {
        String code;
        do {
            System.out.println("Введите код: ");
            Scanner sc = new Scanner(System.in);
            code = sc.next();
            Pattern bin7 = Pattern.compile("^[01]+$");
            Pattern amount = Pattern.compile("^.{7}$");
            Matcher binVer = bin7.matcher(code);
            Matcher amountVer = amount.matcher(code);
            boolean binFormat = binVer.find();
            boolean amountFormat = amountVer.find();
            if (binFormat) {
                if (amountFormat) {
                    char[] input = code.toCharArray();
                    byte[] chArr = new byte[input.length];
                    for (int i = 0; i <= input.length - 1; i++) {
                        chArr[i] = (byte) Character.getNumericValue(input[i]);
                    }
                    boolean[] b = new boolean[chArr.length];
                    for (int i = 0; i < chArr.length; i++) {
                        b[i] = chArr[i] != 0;
                    }
                    boolean s1 = b[0] ^ b[2] ^ b[4] ^ b[5];
                    boolean s2 = b[1] ^ b[2] ^ b[5] ^ b[6];
                    boolean s3 = b[3] ^ b[4] ^ b[5] ^ b[6];
                    int siu = 4 * ((s1) ? 1 : 0) + 2 * ((s2) ? 1 : 0) + ((s3) ? 1 : 0) - 1;
                    if (siu == -1) {
                        System.out.println("Нет ошибок.");
                    } else {
                        b[siu] = !b[siu];
                        int i1f = b[2] ? 1 : 0;
                        int i2f = b[4] ? 1 : 0;
                        int i3f = b[5] ? 1 : 0;
                        int i4f = b[6] ? 1 : 0;
                        System.out.println("Исправленное сообщение: " + "" + i1f + "" + i2f + "" + i3f + "" + i4f);
                    }
                } else {
                    System.out.println("Не забыли ли вы, что данная программа принимает точно и только 7 цифр?");
                }
            } else {
                System.out.println("Зачем творить дичь? Формат ввода - двоичная.");
            }
        } while (!code.equals("stop"));
    }
}

File: Software Engineering/Computer Science/Labs/Lab2/HammingCode.py
import re
while True:
    codeInput = input("Введите код: ")
    if codeInput == "stop":
        break
    else:
        bin7 = re.compile("^[01]+$")
        if re.findall(bin7, codeInput):
            binPlace = re.compile("^[01]{7}$")
            if re.findall(binPlace, codeInput):
                codeList = [int(b) for b in str(codeInput)]
                r1 = bool(codeList[0])
                r2 = bool(codeList[1])
                i1 = bool(codeList[2])
                r3 = bool(codeList[3])
                i2 = bool(codeList[4])
                i3 = bool(codeList[5])
                i4 = bool(codeList[6])
                boolList = [r1, r2, i1, r3, i2, i3, i4]
                s1 = r1 ^ i1 ^ i2 ^ i4
                s2 = r2 ^ i1 ^ i3 ^ i4
                s3 = r3 ^ i2 ^ i3 ^ i4
                s1i = int(s1)
                s2i = int(s2)
                s3i = int(s3)
                notIndex = (s3i * 4 + s2i * 2 + s1i)-1
                if notIndex == -1:
                    print("Нет ошибок. Хорошего дня!")
                else:
                    boolList[notIndex] = not boolList[notIndex]
                    i1f = int(boolList[2])
                    i2f = int(boolList[4])
                    i3f = int(boolList[5])
                    i4f = int(boolList[6])
                    print("Исправленное сообщение: ", i1f, i2f, i3f, i4f)
            else:
                print("Не забыли ли вы, что данная программа принимает точно и только 7 цифр?")
        else:
            print("Зачем творить дичь? Формат ввода - двоичная.")
